tests/common/Language/Java/InlineSpec.hs

```diff
@@ -122,3 +128,27 @@ spec = do
           st :: J ('Class "java.lang.Thread$State") <-
             [java| Thread.State.NEW |]
           [java| $st == Thread.State.NEW |] `shouldReturn` True
+
+      it "Supports modified utf-8 encoding from Haskell to Java" $
+          withLocalRef (reflect ("a\NULb" :: T.Text)) $ \jString ->
+            [java| new String("a\0b").equals($jString) |] `shouldReturn` True
+
+      it "Supports modified utf-8 encoding from Java to Haskell" $ do
+          jString <- [java| new String("a\0b") |]
+          (reify jString) `shouldReturn` ("a\NULb" :: T.Text)
+
+      prop "Processes Unicode code points as the JVM does" $ do
+        \u -> ioProperty $ do
+          let chars :: [Char] = fromUnicode u
+          let text = T.pack chars
+          let codePoints :: [Int32] = map (fromIntegral . ord) chars
```

Comment:
If you use vectors instead of lists, the java type will be `int[]` when reflecting.